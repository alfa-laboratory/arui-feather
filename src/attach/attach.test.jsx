/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

import React from 'react';
import { mount } from 'enzyme';

import Attach from './attach';

describe('attach', () => {
    it('should render without problems', () => {
        let attach = mount(<Attach />);

        expect(attach).toMatchSnapshot();
    });

    it('should set/unset class on attach focused/unfocused', () => {
        const attach = mount(<Attach />);
        const instance = attach.instance();
        const inputRef = instance.input;

        jest.spyOn(inputRef, 'focus');
        jest.spyOn(inputRef, 'blur');

        instance.focus();

        expect(inputRef.focus).toHaveBeenCalledTimes(1);

        instance.blur();

        expect(inputRef.blur).toHaveBeenCalledTimes(1);
    });

    it('should set/unset class on attach hovered/unhovered', () => {
        let attach = mount(<Attach />);

        attach.simulate('mouseEnter');
        expect(attach.getDOMNode().className).toContain('attach_hovered');

        attach.simulate('mouseLeave');
        expect(attach.getDOMNode().className).not.toContain('.attach_hovered');
    });

    it('should call `onClick` callback after attach button was clicked', () => {
        const onClick = jest.fn();
        const attach = mount(<Attach onClick={ onClick } />);

        attach.find('.button').simulate('click');

        expect(onClick).toHaveBeenCalledTimes(1);
    });

    it('should call `onFocus` callback after attach was focused', () => {
        const onFocus = jest.fn();
        const attach = mount(<Attach onFocus={ onFocus } />);

        attach.find('.button').simulate('focus');

        expect(onFocus).toHaveBeenCalledTimes(1);
    });

    it('should call `onBlur` callback after attach was blured', () => {
        const onBlur = jest.fn();
        const attach = mount(<Attach onBlur={ onBlur } />);

        attach.find('.button').simulate('blur');

        expect(onBlur).toHaveBeenCalledTimes(1);
    });

    it('should call `onMouseEnter` callback after attach was hovered', () => {
        const onMouseEnter = jest.fn();
        const attach = mount(<Attach onMouseEnter={ onMouseEnter } />);

        attach.simulate('mouseEnter');

        expect(onMouseEnter).toHaveBeenCalledTimes(1);
    });

    it('should call `onMouseLeave` callback after attach was leaved by cursor', () => {
        const onMouseLeave = jest.fn();
        const attach = mount(<Attach onMouseLeave={ onMouseLeave } />);

        attach.simulate('mouseLeave');

        expect(onMouseLeave).toHaveBeenCalledTimes(1);
    });

    it('should render selected file name if one file selected', () => {
        let attach = mount(<Attach />);
        let controlNode = attach.find('.attach__control');
        controlNode.simulate('change', {
            target: {
                files: [{
                    name: 'test.txt', type: 'application/text'
                }]
            }
        });

        let fileTitleNode = attach.find('.attach__text');
        expect(fileTitleNode.text()).toContain('test.txt');
    });

    it('should render selected files count if several files selected', () => {
        let attach = mount(<Attach />);
        let controlNode = attach.find('.attach__control');
        controlNode.simulate('change', {
            target: {
                files: [
                    {
                        name: 'test1.txt', type: 'application/text'
                    },
                    {
                        name: 'test2.txt', type: 'application/text'
                    },
                    {
                        name: 'test3.txt', type: 'application/text'
                    }
                ]
            }
        });

        let filesTitleNode = attach.find('.attach__text abbr');
        expect(filesTitleNode.text()).toContain('3 файла');
    });

    it('should render "no file" and clear input value after clear button was clicked', () => {
        let attach = mount(<Attach />);
        let controlNode = attach.find('.attach__control');
        controlNode.simulate('change', {
            target: {
                files: [{
                    name: 'test.txt', type: 'application/text'
                }]
            }
        });

        let fileTitleNode = attach.find('.attach__text');
        expect(fileTitleNode.text()).toContain('test.txt');

        let clearButtonNode = attach.find('.attach__clear');
        clearButtonNode.simulate('click');

        let statusNode = attach.find('.attach__no-file');
        expect(statusNode.text()).toContain('Нет файла');
        expect(controlNode.props().value).toBeFalsy();
    });
});
