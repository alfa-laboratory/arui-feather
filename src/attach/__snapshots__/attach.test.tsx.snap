// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`attach should render without problems 1`] = `
<Attach
  buttonContent="Выберите файл"
  disabled={false}
  multiple={false}
  noFileText="Нет файла"
  size="m"
  tabIndex={0}
>
  <span
    className="attach attach_size_m"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <ForwardRef(ThemedComponent(Button))
      className="attach__button"
      disabled={false}
      focused={false}
      icon={
        <ForwardRef(ThemedComponent(IconAttachment))
          size="m"
        />
      }
      leftAddons={
        <label
          className="attach__label"
        >
          <input
            className="attach__control"
            disabled={false}
            multiple={false}
            onChange={[Function]}
            tabIndex={-1}
            type="file"
          />
        </label>
      }
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      size="m"
      tabIndex={0}
      tag="span"
    >
      <Button
        className="attach__button"
        disabled={false}
        focused={false}
        formNoValidate={false}
        icon={
          <ForwardRef(ThemedComponent(IconAttachment))
            size="m"
          />
        }
        leftAddons={
          <label
            className="attach__label"
          >
            <input
              className="attach__control"
              disabled={false}
              multiple={false}
              onChange={[Function]}
              tabIndex={-1}
              type="file"
            />
          </label>
        }
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        size="m"
        tabIndex={0}
        tag="span"
        theme="alfa-on-white"
        type="button"
      >
        <span
          className="button attach__button button_size_m button_theme_alfa-on-white"
          disabled={false}
          formNoValidate={null}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          onMouseOut={[Function]}
          onMouseUp={[Function]}
          role="button"
          tabIndex={0}
          type="button"
        >
          <span
            className="button__addon"
            key="left-addons"
          >
            <label
              className="attach__label"
            >
              <input
                className="attach__control"
                disabled={false}
                multiple={false}
                onChange={[Function]}
                tabIndex={-1}
                type="file"
              />
            </label>
          </span>
          <span
            className="button__content"
            key="content"
          >
            <span
              className="button__icon"
              key="icon"
            >
              <ForwardRef(ThemedComponent(IconAttachment))
                size="m"
              >
                <IconAttachment
                  size="m"
                  theme="alfa-on-white"
                >
                  <ForwardRef(ThemedComponent(Icon))
                    name="attachment"
                    size="m"
                    theme="alfa-on-white"
                  >
                    <Icon
                      name="attachment"
                      size="m"
                      theme="alfa-on-white"
                    >
                      <span
                        className="icon icon_size_m icon_name_attachment icon_theme_alfa-on-white"
                      />
                    </Icon>
                  </ForwardRef(ThemedComponent(Icon))>
                </IconAttachment>
              </ForwardRef(ThemedComponent(IconAttachment))>
            </span>
            <span
              className="button__text"
              key="text"
            >
              Выберите файл
            </span>
          </span>
        </span>
      </Button>
    </ForwardRef(ThemedComponent(Button))>
    <span
      className="attach__no-file"
    >
      Нет файла
    </span>
  </span>
</Attach>
`;
